src/emulator/apphosting/serve.ts

```diff
@@ -36,12 +40,53 @@ export async function start(options?: StartOptions): Promise<{ hostname: string;
     port += 1;
   }
 
-  serve(port, options?.startCommand, options?.rootDirectory);
+  await serve(options?.projectId, port, options?.startCommand, options?.rootDirectory);
 
   return { hostname, port };
 }
 
+// Matches a fully qualified secret or version name, e.g.
+// projects/my-project/secrets/my-secret/versions/1
+// projects/my-project/secrets/my-secret/versions/latest
+// projects/my-project/secrets/my-secret
+const secretResourceRegex =
+  /^projects\/([^/]+)\/secrets\/([^/]+)(?:\/versions\/((?:latest)|\d+))?$/;
+
+// Matches a shorthand for a project-relative secret, with optional version, e.g.
+// my-secret
+// my-secret@1
+// my-secret@latest
+const secretShorthandRegex = /^([^/@]+)(?:@((?:latest)|\d+))?$/;
+
+async function loadSecret(project: string | undefined, name: string): Promise<string> {
+  let projectId: string;
+  let secretId: string;
+  let version: string;
+  const match = secretResourceRegex.exec(name);
+  if (match) {
+    projectId = match[1];
+    secretId = match[2];
+    version = match[3] || "latest";
+  } else {
+    const match = secretShorthandRegex.exec(name);
+    if (!match) {
+      throw new FirebaseError(`Invalid secret name: ${name}`);
+    }
+    if (!project) {
+      throw new FirebaseError(
+        `Cannot load secret ${match[1]} without a project. ` +
+          `Please use ${clc.bold("firebase use")} or pass the --project flag.`,
+      );
+    }
+    projectId = project;
+    secretId = match[1];
+    version = match[2] || "latest";
+  }
+  return await secrets.accessSecretVersion(projectId, secretId, version);
```

Comment:
can we try/catch this to print user friendly log when status == 401 with suggestion to use `firebase apphosting:secrets:grantaccess` command?